test/csuite/wt4156_metadata_salvage/main.c

```diff
@@ -0,0 +1,462 @@
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+#include <sys/wait.h>
+#include <signal.h>
+
+#define	CORRUPT "file:zzz-corrupt.SS"
+#define	KEY	"key"
+#define	VALUE	"value,value,value"
+
+/*
+ * NOTE: This assumes the default page size of 4096. If that changes these
+ * sizes need to change along with it.
+ */
+#define	APP_MD_SIZE 	4096
+#define	APP_BUF_SIZE	3 * 1024
+#define	APP_STR		"long app metadata. "
+
+static bool saw_corruption = false;
+static bool test_abort = false;
+
+static int
+handle_message(WT_EVENT_HANDLER *handler,
+    WT_SESSION *session, int error, const char *message)
+{
+	(void)(handler);
+
+	/* Skip the error messages we're expecting to see. */
+	if ((strstr(message, "database corruption detected") != NULL))
+		saw_corruption = true;
+
+	(void)fprintf(stderr, "%s: %s\n",
+	    message, session->strerror(session, error));
+	if (test_abort) {
+		fprintf(stderr, "Got unexpected error. Aborting\n");
+		abort();
+	}
+	return (0);
+}
+
+static WT_EVENT_HANDLER event_handler = {
+	handle_message,
+	NULL,
+	NULL,
+	NULL
+};
+
+typedef struct table_info {
+	const char *name;
+	const char *kvformat;
+	bool verified;
+} TABLE_INFO;
+
+/*
+ * byte_str --
+ * 	A byte-string version to find a sub-string. The metadata we read
+ * 	contains a lot of zeroes so we cannot use string-based functions.
+ */
+static uint8_t *
+byte_str(uint8_t *buf, size_t bufsize, const char *str)
+{
+	size_t buflen, slen;
+	uint8_t *end, *p, *s;
+	int c;
+
+	p = buf;
+	end = buf + bufsize;
+	s = NULL;
+	c = (int)str[0];
+	buflen = bufsize;
+	slen = strlen(str);
+	/*
+	 * Find the first character and then compare.
+	 */
+	while ((s = memchr(p, c, buflen)) != NULL) {
+		/*
+		 * If we don't have enough buffer left to compare we do not
+		 * have a match.
+		 */
+		buflen = (size_t)(end - s);
+		if (buflen < slen)
+			return (NULL);
+		if (memcmp(s, str, slen) == 0)
+			return (s);
+		/*
+		 * This one didn't match, increment in the buffer and find the
+		 * next one.
+		 */
+		++s;
+		p = s;
+	}
+	return (NULL);
+}
+
+/*
+ * cursor_insert --
+ *	Insert some data into a table.
+ */
+static void
+cursor_insert(WT_SESSION *session, const char *uri)
+{
+	WT_CURSOR *cursor;
+	WT_ITEM vu;
+	char keybuf[100], valuebuf[100];
+	bool recno;
+
+	/* Reserve requires a running transaction. */
+	testutil_check(session->begin_transaction(session, NULL));
```

Comment:
I don't see a matching commit_transaction, and I'm confused about the comment "Reserve requires a running transaction."